usermods/user_fx/user_fx.cpp

```diff
@@ -0,0 +1,116 @@
+#include "wled.h"
+
+// for information how FX metadata strings work see https://kno.wled.ge/interfaces/json-api/#effect-metadata
+
+// static effect, used if an effect fails to initialize
+static uint16_t mode_static(void) {
+  SEGMENT.fill(SEGCOLOR(0));
+  return strip.isOffRefreshRequired() ? FRAMETIME : 350;
+}
+
+/////////////////////////
+//  User FX functions  //
+/////////////////////////
+
+// Diffusion Fire: fire effect intended for 2D setups smaller than 16x16
+static uint16_t mode_diffusionfire(void) {
+  if (!strip.isMatrix || !SEGMENT.is2D())
+    return mode_static();  // not a 2D set-up
+
+  const int cols = SEG_W;
+  const int rows = SEG_H;
+  const auto XY = [&](int x, int y) { return x + y * cols; };
+
+  const uint8_t refresh_hz = map(SEGMENT.speed, 0, 255, 20, 80);
+  const unsigned refresh_ms = 1000 / refresh_hz;
+  const int16_t diffusion = map(SEGMENT.custom1, 0, 255, 0, 100);
+  const uint8_t spark_rate = SEGMENT.intensity;
+  const uint8_t turbulence = SEGMENT.custom2;
+
+  unsigned dataSize = SEGMENT.length(); // allocate persistent data for heat value for each pixel
+  if (!SEGENV.allocateData(dataSize))
+    return mode_static();  // allocation failed
+
+  if (SEGENV.call == 0) {
```

Comment:
I would also add comment like "when `call` is 0 (first execution of effect function) clear segment".